### ISSUE TITLE
"BPF program is too large." on kernel 5.15.0-72, with cilium 1.11.15

### ISSUE THREAD

**Issue by semekh** [cilium/cilium#25500]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

My machine's kernel recently auto-upgraded from 5.15.0-71 to 5.15.0-72 and after that, cilium pods fail to load their bpf programs with the message:

```
level=error msg="endpoint regeneration failed" containerID= datapathPolicyRevision=0 desiredPolicyRevision=658 endpointID=1306 error="Failed to load prog with tc: exit status 1" identity=1836496 ipv4= ipv6= k8sPodName=/ subsys=endpoint
...
level=warning msg="BPF program is too large. Processed 1000001 insn" subsys=datapath-loader
```

### Cilium Version

Client: 1.11.15 c5577e8 2023-03-15T19:19:16+01:00 go version go1.17.13 linux/amd64
Daemon: 1.11.15 c5577e8 2023-03-15T19:19:16+01:00 go version go1.17.13 linux/amd64

### Kernel Version

Linux ... 5.15.0-72-generic #79~20.04.1-Ubuntu SMP Thu Apr 20 22:12:07 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.13", GitCommit:"592eca05be27f7d927d0b25cbb4241d75a9574bf", GitTreeState:"clean", BuildDate:"2022-10-12T10:50:48Z", GoVersion:"go1.17.13", Compiler:"gc", Platform:"linux/amd64"}

### Anything else?

I tried upgrading cilium to 1.11.16 and it didn't resolve the issue.
Downgrading the kernel to 5.15.0-71 fixes the issue.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by phealy** [comment]
This seems to be occurring on kernels that were patched for the [most recent security update by canonical](https://lwn.net/Articles/932125/), as we also hit it on the azure 1038 kernel on Ubuntu 22.04.

**Comment by phealy** [comment]
@semekh a side note, it is not hitting on Cilium 1.12.8, but did hit on 1.10.20.

**Comment by joestringer** [comment]
Broadly three aspects can play a role in issues like this:
- Kernel version
- Cilium version
- Cilium configuration

Regarding the kernel version, it may be worth filing an issue upstream with Canonical to let them know that this kernel has introduced a regression in the BPF subsystem. In general, rolling back the kernel version could be a mitigation for an issue like this, however given that this kernel version addressed some security issues, it would be up to the users in any given environment to weigh up the risk of those security vulnerabilities vs. rolling back in order to mitigate in the short term.

Regarding the Cilium version, given that @phealy reports above that Cilium 1.12.x is not impacted, upgrading may be the most pragmatic approach to be able to receive these kernel security updates and still run Cilium.

There may be possibilities to mitigate such an issue by changing Cilium datapath configuration options, but the exact solution is not known at this time. In the main branch, we maintain a [variety of datapath configurations](https://github.com/cilium/cilium/tree/main/bpf/complexity-tests) to test in order to detect issues like this early in order to try to mitigate them before they reach users. However, these are not prepared in a format that's ready for users to consume. If anyone wants to take a stab at identifying specifically which configurations fail for Cilium 1.11 on this kernel version, it'd be interesting to hear reports of what works and what doesn't. In the mean time, upgrading the Cilium version would be a more reliable approach to resolving this problem.

**Comment by margamanterola** [comment]
This is now affecting our CI for Cilium 1.11: https://github.com/cilium/cilium/pull/25943#issuecomment-1578878257

This kernel commit is the likely culprit: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=354e8f1970f821d4952458f77b1ab6c3eb24d530

Of the changes between Cilium 1.11 and Cilium 1.12 , https://github.com/cilium/cilium/pull/19818 had significant impact in complexity and is likely the reason we are not hitting this issue in Cilium 1.12. However, that's a large refactor that is not easy to backport.

Assigning this issue to Anton, as he's looking into ways of solving this without backporting the above change.

**Comment by aspsk** [comment]
@semekh @phealy Could you please post your cilium agent configuration here? In particular, are you enabling ipsec, tunneling vs. native, dual stack or ipv4-only?

**Comment by aspsk** [comment]
@semekh could you also post if you are running with `kube-proxy-replacement` set to `strict` or to `partial`? (Ideally, could you share the `cilium config -a` from within a cilium pod?)

**Comment by semekh** [comment]
`kube-proxy-replacement` is set to `strict` mode. Also here's the output of `cilium config -a`:

```
#### Read-only configurations ####
APIRateLimit
ARPPingKernelManaged              : false
ARPPingRefreshPeriod              : 30000000000
AgentHealthPort                   : 9879
AgentLabels                       : []
AgentNotReadyNodeTaintKey         : node.cilium.io/agent-not-ready
AllocatorListTimeout              : 180000000000
AllowICMPFragNeeded               : true
AllowLocalhost                    : always
AnnotateK8sNode                   : true
AutoCreateCiliumNodeResource      : true
BGPAnnounceLBIP                   : false
BGPAnnouncePodCIDR                : false
BGPConfigPath                     : /var/lib/cilium/bgp/config.yaml
BPFCompileDebug                   : 
BPFMapsDynamicSizeRatio           : 0.0025
BPFRoot                           : 
BPFSocketLBHostnsOnly             : true
BpfDir                            : /var/lib/cilium/bpf
BypassIPAvailabilityUponRestore   : false
CGroupRoot                        : /run/cilium/cgroupv2
CNIChainingMode                   : 
CRDWaitTimeout                    : 300000000000
CTMapEntriesGlobalAny             : 277079
CTMapEntriesGlobalTCP             : 554159
CTMapEntriesTimeoutAny            : 60000000000
CTMapEntriesTimeoutFIN            : 10000000000
CTMapEntriesTimeoutSVCAny         : 60000000000
CTMapEntriesTimeoutSVCTCP         : 21600000000000
CTMapEntriesTimeoutSYN            : 60000000000
CTMapEntriesTimeoutTCP            : 21600000000000
CertDirectory                     : /var/run/cilium/certs
CgroupPathMKE                     : 
ClockSource                       : 1
ClusterHealthPort                 : 4240
ClusterID                         : 28
ClusterMeshConfig                 : /var/lib/cilium/clustermesh/
CompilerFlags                     : []
ConfigDir                         : /tmp/cilium/config-map
ConfigFile                        : 
ConntrackGCInterval               : 0
DNSMaxIPsPerRestoredRule          : 1000
DNSProxyConcurrencyLimit          : 0
DNSProxyConcurrencyProcessingGracePeriod: 0
DatapathMode                      : veth
Debug                             : false
DebugVerbose                      : []
DeriveMasqIPAddrFromDevice        : 
DevicePreFilter                   : undefined
DisableCNPStatusUpdates           : true
DisableCiliumEndpointCRD          : false
DisableConntrack                  : false
DisableEnvoyVersionCheck          : true
DisableIptablesFeederRules        : []
DryMode                           : false
EgressMasqueradeInterfaces        : 
EgressMultiHomeIPRuleCompat       : false
EnableAutoDirectRouting           : false
EnableAutoProtectNodePortRange    : true
EnableBPFClockProbe               : true
EnableBPFMasquerade               : false
EnableBPFTProxy                   : false
EnableBandwidthManager            : false
EnableCiliumEndpointSlice         : false
EnableCustomCalls                 : false
EnableEndpointHealthChecking      : true
EnableEndpointRoutes              : false
EnableExternalIPs                 : true
EnableHealthCheckNodePort         : true
EnableHealthChecking              : true
EnableHealthDatapath              : false
EnableHostFirewall                : false
EnableHostIPRestore               : true
EnableHostLegacyRouting           : true
EnableHostPort                    : true
EnableHostReachableServices       : true
EnableHostServicesPeer            : true
EnableHostServicesTCP             : true
EnableHostServicesUDP             : true
EnableHubble                      : false
EnableHubbleRecorderAPI           : true
EnableICMPRules                   : false
EnableIPMasqAgent                 : false
EnableIPSec                       : false
EnableIPv4                        : true
EnableIPv4EgressGateway           : false
EnableIPv4FragmentsTracking       : true
EnableIPv4Masquerade              : true
EnableIPv6                        : false
EnableIPv6Masquerade              : false
EnableIPv6NDP                     : false
EnableIdentityMark                : true
EnableK8sTerminatingEndpoint      : true
EnableL2NeighDiscovery            : true
EnableL7Proxy                     : false
EnableLocalNodeRoute              : true
EnableLocalRedirectPolicy         : false
EnableMKE                         : false
EnableMonitor                     : true
EnableNodePort                    : true
EnablePolicy                      : default
EnableRecorder                    : false
EnableRemoteNodeIdentity          : true
EnableSVCSourceRangeCheck         : true
EnableServiceTopology             : false
EnableSessionAffinity             : true
EnableStaleCiliumEndpointCleanup  : true
EnableTracing                     : false
EnableWellKnownIdentities         : false
EnableWireguard                   : false
EnableWireguardUserspaceFallback  : false
EnableXDPPrefilter                : false
EnableXTSocketFallback            : true
EncryptInterface                  : []
EncryptNode                       : false
EndpointGCInterval                : 300000000000
EndpointInterfaceNamePrefix       : lxc+
EndpointQueueSize                 : 25
EndpointStatus
EnvoyLog                          : 
EnvoyLogPath                      : 
ExcludeLocalAddresses             : <nil>
ExternalClusterIP                 : false
FQDNProxyResponseMaxDelay         : 100000000
FQDNRegexCompileLRUSize           : 1024
FQDNRejectResponse                : refused
FixedIdentityMapping
ForceLocalPolicyEvalAtSource      : true
FragmentsMapEntries               : 8192
HTTP403Message                    : 
HTTPIdleTimeout                   : 0
HTTPMaxGRPCTimeout                : 0
HTTPNormalizePath                 : true
HTTPRequestTimeout                : 3600
HTTPRetryCount                    : 3
HTTPRetryTimeout                  : 0
HostDevice                        : cilium_host
HostV4Addr                        : 
HostV6Addr                        : 
HubbleEventBufferCapacity         : 4095
HubbleEventQueueSize              : 12288
HubbleExportFileCompress          : false
HubbleExportFileMaxBackups        : 5
HubbleExportFileMaxSizeMB         : 10
HubbleExportFilePath              : 
HubbleListenAddress               : 
HubbleMetrics                     : []
HubbleMetricsServer               : 
HubbleRecorderSinkQueueSize       : 1024
HubbleRecorderStoragePath         : /var/run/cilium/pcaps
HubbleSocketPath                  : /var/run/cilium/hubble.sock
HubbleTLSCertFile                 : 
HubbleTLSClientCAFiles            : []
HubbleTLSDisabled                 : false
HubbleTLSKeyFile                  : 
IPAM                              : cluster-pool
IPAllocationTimeout               : 120000000000
IPMasqAgentConfigPath             : /etc/config/ip-masq-agent
IPSecKeyFile                      : 
IPTablesLockTimeout               : 5000000000
IPTablesRandomFully               : false
IPv4NativeRoutingCIDR             : <nil>
IPv4NodeAddr                      : auto
IPv4PodSubnets                    : []
IPv4Range                         : auto
IPv4ServiceRange                  : auto
IPv6ClusterAllocCIDR              : f00d::/64
IPv6ClusterAllocCIDRBase          : f00d::
IPv6MCastDevice                   : 
IPv6NodeAddr                      : auto
IPv6PodSubnets                    : []
IPv6Range                         : auto
IPv6ServiceRange                  : auto
IdentityAllocationMode            : crd
IdentityChangeGracePeriod         : 5000000000
IdentityRestoreGracePeriod        : 600000000000
InstallIptRules                   : true
InstallNoConntrackIptRules        : false
Ipvlan
        MasterDeviceIndex         : Disabled
        OperationMode             : 
IpvlanMasterDevice                : undefined
JoinCluster                       : false
K8sAPIServer                      : 
K8sClientBurst                    : 0
K8sClientQPSLimit                 : 0
K8sEnableAPIDiscovery             : false
K8sEnableK8sEndpointSlice         : true
K8sEnableLeasesFallbackDiscovery  : false
K8sEventHandover                  : false
K8sHeartbeatTimeout               : 30000000000
K8sKubeConfigPath                 : 
K8sNamespace                      : kube-system
K8sRequireIPv4PodCIDR             : true
K8sRequireIPv6PodCIDR             : false
K8sServiceCacheSize               : 128
K8sServiceProxyName               : 
K8sSyncTimeout                    : 180000000000
K8sWatcherEndpointSelector        : metadata.name!=kube-scheduler,metadata.name!=kube-controller-manager,metadata.name!=etcd-operator,metadata.name!=gcp-controller-manager
KVStore                           : 
KVStoreOpt
KVstoreConnectivityTimeout        : 120000000000
KVstoreKeepAliveInterval          : 300000000000
KVstoreLeaseTTL                   : 900000000000
KVstoreMaxConsecutiveQuorumErrors : 2
KVstorePeriodicSync               : 300000000000
KeepConfig                        : false
KernelHz                          : 250
KubeProxyReplacement              : strict
KubeProxyReplacementHealthzBindAddr: 0.0.0.0:10256
LBDevInheritIPAddr                : 
LBMapEntries                      : 65536
LabelPrefixFile                   : 
Labels                            : []
LibDir                            : /var/lib/cilium
LoadBalancerDSRDispatch           : opt
LoadBalancerDSRL4Xlate            : frontend
LoadBalancerPMTUDiscovery         : false
LoadBalancerPreserveWorldID       : false
LoadBalancerRSSv4
        IP                        : 
        Mask                      : <nil>
LoadBalancerRSSv4CIDR             : 
LoadBalancerRSSv6
        IP                        : 
        Mask                      : <nil>
LoadBalancerRSSv6CIDR             : 
LocalRouterIPv4                   : 
LocalRouterIPv6                   : 
LogDriver                         : []
LogOpt
LogSystemLoadConfig               : false
Logstash                          : false
LoopbackIPv4                      : 169.254.42.1
MTU                               : 0
MaglevTableSize                   : 16381
MaxControllerInterval             : 0
MetricsConfig
        APIInteractionsEnabled    : Enabled
        APILimiterAdjustmentFactor: Enabled
        APILimiterProcessedRequests: Enabled
        APILimiterProcessingDuration: Enabled
        APILimiterRateLimit       : Enabled
        APILimiterRequestsInFlight: Enabled
        APILimiterWaitDuration    : Enabled
        APILimiterWaitHistoryDuration: Disabled
        ArpingRequestsTotalEnabled: Disabled
        BPFMapOps                 : Enabled
        BPFMapPressure            : Disabled
        BPFSyscallDurationEnabled : Disabled
        ConntrackDumpResetsEnabled: Enabled
        ConntrackGCDurationEnabled: Enabled
        ConntrackGCKeyFallbacksEnabled: Enabled
        ConntrackGCRunsEnabled    : Enabled
        ConntrackGCSizeEnabled    : Enabled
        ControllerRunsDurationEnabled: Enabled
        ControllerRunsEnabled     : Enabled
        DropBytesEnabled          : Enabled
        DropCountEnabled          : Enabled
        EndpointPropagationDelayEnabled: Enabled
        EndpointRegenerationCountEnabled: Enabled
        EndpointRegenerationTimeStatsEnabled: Enabled
        EndpointStateCountEnabled : Enabled
        ErrorsWarningsEnabled     : Enabled
        EventLagK8sEnabled        : Enabled
        EventTSAPIEnabled         : Enabled
        EventTSContainerdEnabled  : Enabled
        EventTSK8sEnabled         : Enabled
        FQDNActiveIPs             : Disabled
        FQDNActiveNames           : Disabled
        FQDNActiveZombiesConnections: Disabled
        FQDNGarbageCollectorCleanedTotalEnabled: Enabled
        FQDNSemaphoreRejectedTotal: Disabled
        ForwardBytesEnabled       : Enabled
        IPCacheErrorsTotalEnabled : Enabled
        IPCacheEventsTotalEnabled : Disabled
        IdentityCountEnabled      : Enabled
        IpamEventEnabled          : Enabled
        KVStoreEventsQueueDurationEnabled: Enabled
        KVStoreOperationsDurationEnabled: Enabled
        KVStoreQuorumErrorsEnabled: Enabled
        KubernetesAPICallsEnabled : Enabled
        KubernetesAPIInteractionsEnabled: Enabled
        KubernetesCNPStatusCompletionEnabled: Enabled
        KubernetesEventProcessedEnabled: Enabled
        KubernetesEventReceivedEnabled: Enabled
        KubernetesTerminatingEndpointsEnabled: Enabled
        NoOpCounterVecEnabled     : Enabled
        NoOpObserverVecEnabled    : Enabled
        NodeConnectivityLatencyEnabled: Enabled
        NodeConnectivityStatusEnabled: Enabled
        PolicyCountEnabled        : Enabled
        PolicyEndpointStatusEnabled: Enabled
        PolicyImplementationDelayEnabled: Enabled
        PolicyImportErrorsEnabled : Enabled
        PolicyRegenerationCountEnabled: Enabled
        PolicyRegenerationTimeStatsEnabled: Enabled
        PolicyRevisionEnabled     : Disabled
        ProxyDatapathUpdateTimeoutEnabled: Disabled
        ProxyDeniedEnabled        : Enabled
        ProxyForwardedEnabled     : Enabled
        ProxyParseErrorsEnabled   : Enabled
        ProxyPolicyL7Enabled      : Enabled
        ProxyReceivedEnabled      : Enabled
        ProxyRedirectsEnabled     : Enabled
        ServicesCountEnabled      : Enabled
        SignalsHandledEnabled     : Enabled
        SubprocessStartEnabled    : Enabled
        TriggerPolicyUpdateCallDuration: Enabled
        TriggerPolicyUpdateFolds  : Enabled
        TriggerPolicyUpdateTotal  : Enabled
        VersionMetric             : Enabled
ModePreFilter                     : native
Monitor
        cpus                      : 12
        npages                    : 64
        pagesize                  : 4096
MonitorAggregation                : medium
MonitorAggregationFlags           : 255
MonitorAggregationInterval        : 5000000000
MonitorQueueSize                  : 12288
NAT46Prefix
        IP                        : 0.0.0.0
        Mask                      : ////////////////AAAAAA==
NAT46Range                        : 0:0:0:0:0:FFFF::/96
NATMapEntriesGlobal               : 554159
NeighMapEntriesGlobal             : 554159
NodePortAcceleration              : disabled
NodePortAlg                       : random
NodePortBindProtection            : true
NodePortMax                       : 32767
NodePortMin                       : 30000
NodePortMode                      : snat
PProf                             : false
PProfPort                         : 6060
PolicyAuditMode                   : false
PolicyMapEntries                  : 16384
PolicyQueueSize                   : 100
PolicyTriggerInterval             : 1000000000
PreAllocateMaps                   : false
PrependIptablesChains             : true
PrometheusServeAddr               : 
ProxyConnectTimeout               : 1
ProxyGID                          : 1337
ProxyPrometheusPort               : 0
ReadCNIConfiguration              : 
ResetQueueMapping                 : true
RestoreState                      : true
RouteMetric                       : 0
RunDir                            : /var/run/cilium
RunMonitorAgent                   : true
SelectiveRegeneration             : true
SidecarIstioProxyImage            : cilium/istio_proxy
SizeofCTElement                   : 94
SizeofNATElement                  : 94
SizeofNeighElement                : 24
SizeofSockRevElement              : 52
SockRevNatEntries                 : 277079
SocketPath                        : /var/run/cilium/cilium.sock
SockopsEnable                     : false
StateDir                          : /var/run/cilium/state
ToFQDNsEnableDNSCompression       : true
ToFQDNsIdleConnectionGracePeriod  : 0
ToFQDNsMaxDeferredConnectionDeletes: 10000
ToFQDNsMaxIPsPerHost              : 50
ToFQDNsMinTTL                     : 3600
ToFQDNsPreCache                   : 
ToFQDNsProxyPort                  : 0
TracePayloadlen                   : 128
Tunnel                            : vxlan
TunnelPort                        : 8472
UseSingleClusterRoute             : false
VLANBPFBypass                     : []
Version                           : false
WriteCNIConfigurationWhenReady    : 
XDPMode                           : disabled
k8s-configuration                 : 
k8s-endpoint                      : 
##### Read-write configurations #####
Conntrack                         : Enabled
ConntrackAccounting               : Enabled
ConntrackLocal                    : Disabled
Debug                             : Disabled
DebugLB                           : Disabled
DropNotification                  : Enabled
MonitorAggregationLevel           : Medium
PolicyAuditMode                   : Disabled
PolicyTracing                     : Disabled
PolicyVerdictNotification         : Enabled
TraceNotification                 : Enabled
PolicyEnforcement                 : default
```

**Comment by aspsk** [comment]
Thanks @semekh !

**Comment by raxod502-plaid** [comment]
We deployed Cilium 1.11.18 (upgraded from 1.11.10) in a Kubernetes 1.20 cluster and can confirm that it is compatible with newer kernel versions where the old version was not.

**Comment by margamanterola** [comment]
Alright, then. Marking this as fixed on 1.11.18. Thanks for confirming!
